File: Stack/1096.Brace Expansion II.cpp
#include <vector>
#include <set>
#include <string>
using namespace std;

class Solution
{
public:
    vector<char> ops;
    vector<set<string>> setStack;

    void operate()
    {
        int start = setStack.size() - 2;
        int end = setStack.size() - 1;

        if (ops.back() == '+')
        {
            setStack[start].merge(setStack[end]);
        }
        else
        {
            set<string> temp;

            for (auto &item1 : setStack[start])
            {
                for (auto &item2 : setStack[end])
                {
                    temp.insert(item1 + item2);
                }
            }
            setStack[start] = temp;
        }
        ops.pop_back();
        setStack.pop_back();
    }

    vector<string> braceExpansionII(string expression)
    {
        int n = expression.length();

        for (int i = 0; i < n; i++)
        {
            char ch = expression[i];
            if (ch == ',')
            {
                while (ops.size() > 0 && ops.back() == '*')
                {
                    operate();
                }
                ops.push_back('+');
            }
            else if (ch == '{')
            {
                if (i > 0 &&
                    (expression[i - 1] == '}' || isalpha(expression[i - 1])))
                {
                    ops.push_back('*');
                }
                ops.push_back('{');
            }
            else if (ch == '}')
            {
                while (ops.size() && ops.back() != '{')
                {
                    operate();
                }
                ops.pop_back();
            }
            else
            {
                if (i > 0 &&
                    (expression[i - 1] == '}' || isalpha(expression[i - 1])))
                {
                    ops.push_back('*');
                }
                setStack.push_back({string(1, expression[i])});
            }
        }

        while (ops.size() > 0)
        {
            operate();
        }

        return {setStack.back().begin(), setStack.back().end()};
    }
};

File: Stack/1096.Brace Expansion II.ts
function braceExpansionII(expression: string): string[] {
    let opsStack: string[] = [];
    let setStack: Set<string>[] = [];
    let n = expression.length;

    function operate() {
        let start = setStack.length - 2
        let end = setStack.length - 1;

        if (opsStack[opsStack.length - 1] == "+") {
            for (let item of setStack[end]) {
                setStack[start].add(item);
            }
        } else { // "*"
            let temp = new Set<string>();

            for (let item1 of setStack[start]) {
                for (let item2 of setStack[end]) {
                    temp.add(item1 + item2);
                }
            }
            setStack[start] = temp;
        }
        opsStack.pop();
        setStack.pop();
    }

    for (let i = 0; i < n; i++) {
        const char = expression[i];

        if (char == ",") {
            while (opsStack.length && opsStack[opsStack.length - 1] == "*") {
                operate();
            }

            opsStack.push("+");
        } else if (char == "{") {      // opening bracket
            if (i > 0 && (expression[i - 1] == "}" || /[a-z]/.test(expression[i - 1]))) {
                opsStack.push("*");
            }
            opsStack.push("{");
        } else if (char == "}") {               // closing bracket
            while (opsStack.length && opsStack[opsStack.length - 1] !== "{") {
                operate();
            }
            opsStack.pop();
        } else { // character
            if (i > 0 && (expression[i - 1] == "}" || /[a-z]/.test(expression[i - 1]))) {
                opsStack.push("*");
            }
            setStack.push(new Set([char]));
        }
    }

    while (opsStack.length > 0) {
        operate();
    }

    return Array.from(setStack[setStack.length - 1]).sort();
};